pkg/coscheduling/core/core.go

```diff
@@ -0,0 +1,534 @@
+	preAllocatedResource := getPreAllocatedResource(maxSchedulingPGStatus, preScheduled)
+	preAllocatedResource.Add(getPodResourceRequire(pod).ResourceList())
+	if !pgMgr.compareClusterResourceAndRequire(maxSchedulingPGStatus.Pod, &preAllocatedResource, pgMgr.reserveResourcePercentage) {
+		errMsg := fmt.Sprintf("cluster resource not enough, preAllocated: %v, current requeired: %v",
+			preAllocatedResource, *pgs.PodGroup.Spec.MinResources)
+		// Check priority
+		if podutil.GetPodPriority(pod) <= podutil.GetPodPriority(maxSchedulingPGStatus.Pod) {
+			pgMgr.AddToDenyCache(pgFullName)
+			return fmt.Errorf(errMsg)
+		}
+		evictionList, _ := pgMgr.simulateEvictGroup(pgs, pgFullName)
+		klog.V(4).Infof("Eviction list %+v", evictionList)
+		if len(evictionList) > 0 {
+			pgMgr.evictGroup(evictionList)
+			pgMgr.lastDeniedPG.Delete(pgFullName)
+		}
+	}
+	return nil
+}
+
+// PreemptRemovePod requires:
+// 1. group pod could not preempt no-group pod
+// 2. group pod could preempt not running group pod
+// 3. no-group pod preempt group pod
+// 4. no-group pod preempt no-group pod
+func (pgMgr *PodGroupManager) PreemptRemovePod(ctx context.Context, podToSchedule, podToRemove *corev1.Pod) error {
+	klog.Infof("podToSchedule pod %v, podToRemove pod %v", podToSchedule.Name, podToRemove.Name)
+
+	pgNameToRemove, groupToRemove, _ := pgMgr.GetPodGroup(podToRemove)
+	pgNameToSchedule, groupToSchedule, _ := pgMgr.GetPodGroup(podToSchedule)
+
+	// 1. group pod could preempt no-group pod
+	// 2. group pod could preempt fewer scheduling and not running group pod
+
+	// no-group pod preempt no-group pod
+	if groupToRemove == nil && groupToSchedule == nil {
+		return nil
+	}
+
+	// group pod preempt no-group pod: forbid
+	if groupToRemove == nil && groupToSchedule != nil {
+		return fmt.Errorf("group pod %v are forbidden to preempt no-group pod %v", podToSchedule.Name, podToRemove.Name)
+	}
+
+	// no-group pod preempt group pod
+	if groupToRemove != nil && groupToSchedule == nil {
+		return nil
+	}
+
+	// group pod preempt group pod
+	if pgNameToRemove == pgNameToSchedule {
+		return fmt.Errorf("podToSchedule and podToRemove belong to same pod group, do not preempt")
+	}
+
+	// do not allow preempt scheduled group
+	if groupToRemove.PodGroup.Status.Phase == v1alpha1.PodGroupScheduled ||
+		groupToRemove.PodGroup.Status.Phase == v1alpha1.PodGroupRunning || groupToRemove.PodGroup.Status.Scheduled >= groupToRemove.PodGroup.Spec.MinMember {
+		return fmt.Errorf("pod %v belongs to Scheduled or Running pod group can not be preempted", podToRemove.Name)
+	}
+
+	klog.V(5).Infof("Try to preempt pod %v/%v, schedule pod %v/%v", podToRemove.Namespace, podToRemove.Name,
+		podToSchedule.Namespace, podToSchedule.Name)
+	return nil
+}
+
+// Permit permits a pod to run, if the minMember match, it would send a signal to chan.
+func (pgMgr *PodGroupManager) Permit(ctx context.Context, pod *corev1.Pod, nodeName string) (bool, string, error) {
```

Comment:
It looks to me the returned args can be simplfied:

- the key info we want to return is whether it's ready to schedule the whole PodGroup - here you tell this by checking `if err == nil`
- pgFullName shouldn't be returned as an argument as it can be naturally deduced from a given pod at the caller side

So I'd perfer to define the args as `bool, framework.Status`, then in caller side, return the framework.Status if the bool arg is false; otherwise continue the logic.

Reply:
I would not return `framework.Status` in the `pgMgr`. But pgFullName can be removed.